### ISSUE TITLE
 AttributeError: 'module' object has no attribute 'ssl_' on Ubuntu 16.04.1 LTS

### ISSUE THREAD

**Issue by DigNative** [mozilla-services/syncserver#80]
Hi,

some days ago, I upgraded one of my servers from Ubuntu 14.04 LTS to Ubuntu 16.04.1 LTS and noticed that the Firefox Sync Server is not working anymore, throwing errors similar to those reported in issue #57. I completely reinstalled the Sync Server by cloning a fresh copy of the master to a local working dir and running `make build`. A subsequent `make test` shows no errors. I got rid of the previous error.

However, running the server with Apache and the following Apache configuration

```
WSGIProcessGroup sync
WSGIDaemonProcess sync user=www-data group=www-data processes=2 threads=25 python-path=/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/
WSGIPassAuthorization On
WSGIScriptAlias     /       /srv/software/mozilla-syncserver/syncserver.wsgi
```

gives me the following errors in the web server's log:

```
[wsgi:error] mod_wsgi (pid=30328): Target WSGI script '/srv/software/mozilla-syncserver/syncserver.wsgi' cannot be loaded as Python module.
[wsgi:error] mod_wsgi (pid=30328): Exception occurred processing WSGI script '/srv/software/mozilla-syncserver/syncserver.wsgi'.
[wsgi:error] Traceback (most recent call last):
[wsgi:error]   File "/srv/software/mozilla-syncserver/syncserver.wsgi", line 40, in <module>
[wsgi:error]     application = loadapp('config:%s'% ini_file)
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 247, in loadapp
[wsgi:error]     return loadobj(APP, uri, name=name, **kw)
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 271, in loadobj
[wsgi:error]     global_conf=global_conf)
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 296, in loadcontext
[wsgi:error]     global_conf=global_conf)
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 320, in _loadconfig
[wsgi:error]     return loader.get_context(object_type, name, global_conf)
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 454, in get_context
[wsgi:error]     section)
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 476, in _context_from_use
[wsgi:error]     object_type, name=use, global_conf=global_conf)
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 406, in get_context
[wsgi:error]     global_conf=global_conf)
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 296, in loadcontext
[wsgi:error]     global_conf=global_conf)
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 328, in _loadegg
[wsgi:error]     return loader.get_context(object_type, name, global_conf)
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 620, in get_context
[wsgi:error]     object_type, name=name)
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 646, in find_egg_entry_point
[wsgi:error]     possible.append((entry.load(), protocol, entry.name))
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/pkg_resources.py", line 2048, in load
[wsgi:error]   File "/srv/software/mozilla-syncserver/syncserver/__init__.py", line 13, in <module>
[wsgi:error]     import requests.packages.urllib3.contrib.pyopenssl
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/requests/__init__.py", line 53, in <module>
[wsgi:error]     from .packages.urllib3.contrib import pyopenssl
[wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 86, in <module>
[wsgi:error]     DEFAULT_SSL_CIPHER_LIST = util.ssl_.DEFAULT_CIPHERS
[wsgi:error] AttributeError: 'module' object has no attribute 'ssl_'
```

**Comment by rfk** [comment]
Looks like this may be an issue with the `requests` dependency.  I've updated it to the latest release version, could you please try pulling the latest `syncserver` from github and see if it resolves the issue?

**Comment by DigNative** [comment]
Thank you very much for your answer and the modification! I have cloned a fresh copy from the GitHub master and did the steps described above again. Now I am receiving a different kind of error message:

    [wsgi:error] mod_wsgi (pid=26867): Target WSGI script '/srv/software/mozilla-syncserver/syncserver.wsgi' cannot be loaded as Python module.
    [wsgi:error] mod_wsgi (pid=26867): Exception occurred processing WSGI script '/srv/software/mozilla-syncserver/syncserver.wsgi'.
    [wsgi:error] Traceback (most recent call last):
    [wsgi:error]   File "/srv/software/mozilla-syncserver/syncserver.wsgi", line 40, in <module>
    [wsgi:error]     application = loadapp('config:%s'% ini_file)
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 247, in loadapp
    [wsgi:error]     return loadobj(APP, uri, name=name, **kw)
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 271, in loadobj
    [wsgi:error]     global_conf=global_conf)
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 296, in loadcontext
    [wsgi:error]     global_conf=global_conf)
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 320, in _loadconfig
    [wsgi:error]     return loader.get_context(object_type, name, global_conf)
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 454, in get_context
    [wsgi:error]     section)
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 476, in _context_from_use
    [wsgi:error]     object_type, name=use, global_conf=global_conf)
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 406, in get_context
    [wsgi:error]     global_conf=global_conf)
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 296, in loadcontext
    [wsgi:error]     global_conf=global_conf)
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 328, in _loadegg
    [wsgi:error]     return loader.get_context(object_type, name, global_conf)
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 620, in get_context
    [wsgi:error]     object_type, name=name)
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 640, in find_egg_entry_point
    [wsgi:error]     pkg_resources.require(self.spec)
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 943, in require
    [wsgi:error]     needed = self.resolve(parse_requirements(requirements))
    [wsgi:error]   File "/srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 829, in resolve
    [wsgi:error]     raise DistributionNotFound(req, requirers)
    [wsgi:error] DistributionNotFound: The 'syncserver' distribution was not found and is required by the application

**Comment by rfk** [comment]
IIRC this error can occur when mod_wsgi is not configured to correctly load the virtualenv into which the server is installed.  Have you used to `python-path` directive as specified at [1] and pointed it to your own local install?

[1] https://docs.services.mozilla.com/howtos/run-sync-1.5.html#apache-mod-wsgi

**Comment by DigNative** [comment]
Yes, I did (see my first post for the Apache virtual host configuration). I also double-checked the path:

    $ ls -alh /srv/software/mozilla-syncserver/local/lib/python2.7/site-packages/
    total 872K
    drwxrwxr-x 83 www-data www-data 4.0K Aug 13 10:04 .
    drwxrwxr-x  4 www-data www-data 4.0K Aug 13 10:00 ..
    drwxrwxr-x  3 www-data www-data 4.0K Aug 13 10:02 backports
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 beautifulsoup4-4.5.1.dist-info
    drwxrwxr-x  5 www-data www-data 4.0K Aug 13 10:02 browserid
    drwxrwxr-x  4 www-data www-data 4.0K Aug 13 10:02 bs4
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 configparser-3.5.0b2.dist-info
    -rw-rw-r--  1 www-data www-data  314 Aug 13 10:02 configparser-3.5.0b2-py2.7-nspkg.pth
    -rw-rw-r--  1 www-data www-data  854 Aug 13 10:02 configparser.py
    -rw-rw-r--  1 www-data www-data 1.3K Aug 13 10:02 configparser.pyc
    drwxrwxr-x  5 www-data www-data 4.0K Aug 13 10:02 cornice
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 cornice-0.16.2.dist-info
    -rw-rw-r--  1 www-data www-data   37 Aug 13 10:03 easy-install.pth
    -rw-rw-r--  1 www-data www-data  126 Aug 13 10:00 easy_install.py
    -rw-rw-r--  1 www-data www-data  315 Aug 13 10:00 easy_install.pyc
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:04 enum
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:04 enum34-1.1.6.dist-info
    drwxrwxr-x  7 www-data www-data 4.0K Aug 13 10:04 flake8
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:04 flake8-3.0.4.dist-info
    drwxrwxr-x  7 www-data www-data 4.0K Aug 13 10:02 gunicorn
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 gunicorn-19.1.1.dist-info
    drwxrwxr-x  3 www-data www-data 4.0K Aug 13 10:02 hawkauthlib
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 hawkauthlib-0.1.1.dist-info
    drwxrwxr-x  3 www-data www-data 4.0K Aug 13 10:02 konfig
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 konfig-0.9.dist-info
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:04 mccabe-0.5.2.dist-info
    -rw-rw-r--  1 www-data www-data  11K Aug 13 10:04 mccabe.py
    -rw-rw-r--  1 www-data www-data  13K Aug 13 10:04 mccabe.pyc
    drwxrwxr-x  5 www-data www-data 4.0K Aug 13 10:02 mozsvc
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 mozsvc-0.9.dev0-py2.7.egg-info
    drwxrwxr-x  6 www-data www-data 4.0K Aug 13 10:04 nose
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:04 nose-1.3.7.dist-info
    drwxrwxr-x  9 www-data www-data 4.0K Aug 13 10:02 paste
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 Paste-2.0.3.dist-info
    -rw-rw-r--  1 www-data www-data  302 Aug 13 10:02 Paste-2.0.3-py2.7-nspkg.pth
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 PasteDeploy-1.5.2.dist-info
    -rw-rw-r--  1 www-data www-data  304 Aug 13 10:02 PasteDeploy-1.5.2-py2.6-nspkg.pth
    drwxrwxr-x 10 www-data www-data 4.0K Aug 13 10:00 pip
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:00 pip-8.1.2.dist-info
    drwxrwxr-x  4 www-data www-data 4.0K Aug 13 10:00 pkg_resources
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 PyBrowserID-0.11.0.dist-info
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:04 pycodestyle-2.0.0.dist-info
    -rw-rw-r--  1 www-data www-data  82K Aug 13 10:04 pycodestyle.py
    -rw-rw-r--  1 www-data www-data  72K Aug 13 10:04 pycodestyle.pyc
    drwxrwxr-x  4 www-data www-data 4.0K Aug 13 10:04 pyflakes
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:04 pyflakes-1.2.3.dist-info
    drwxrwxr-x  4 www-data www-data 4.0K Aug 13 10:02 pymysql
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 PyMySQL-0.7.6.dist-info
    drwxrwxr-x  3 www-data www-data 4.0K Aug 13 10:02 pymysql_sa
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 pymysql_sa-1.0.dist-info
    drwxrwxr-x  6 www-data www-data 4.0K Aug 13 10:02 pyramid
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 pyramid-1.5.dist-info
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 pyramid_hawkauth
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 pyramid_hawkauth-0.1.0.dist-info
    drwxrwxr-x  3 www-data www-data 4.0K Aug 13 10:02 repoze
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 repoze.lru-0.6.dist-info
    -rw-rw-r--  1 www-data www-data  305 Aug 13 10:02 repoze.lru-0.6-py2.7-nspkg.pth
    drwxrwxr-x  3 www-data www-data 4.0K Aug 13 10:02 requests
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 requests-2.10.0.dist-info
    drwxrwxr-x  4 www-data www-data 4.0K Aug 13 10:00 setuptools
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:00 setuptools-25.2.0.dist-info
    drwxrwxr-x  3 www-data www-data 4.0K Aug 13 10:02 simplejson
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 simplejson-3.4.0.dist-info
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 six-1.10.0.dist-info
    -rw-rw-r--  1 www-data www-data  30K Aug 13 10:02 six.py
    -rw-rw-r--  1 www-data www-data  29K Aug 13 10:02 six.pyc
    drwxrwxr-x 12 www-data www-data 4.0K Aug 13 10:02 sqlalchemy
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 SQLAlchemy-0.9.4.dist-info
    -rw-rw-r--  1 www-data www-data   38 Aug 13 10:03 syncserver.egg-link
    drwxrwxr-x  6 www-data www-data 4.0K Aug 13 10:03 syncstorage
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:03 SyncStorage-1.5.14-py2.7.egg-info
    drwxrwxr-x  3 www-data www-data 4.0K Aug 13 10:02 testfixtures
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 testfixtures-4.10.0.dist-info
    drwxrwxr-x  3 www-data www-data 4.0K Aug 13 10:02 tokenlib
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 tokenlib-0.3.1.dist-info
    drwxrwxr-x  5 www-data www-data 4.0K Aug 13 10:02 tokenserver
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 tokenserver-1.2.23-py2.7.egg-info
    drwxrwxr-x  3 www-data www-data 4.0K Aug 13 10:02 translationstring
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 translationstring-1.3.dist-info
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 umemcache-1.6.3.dist-info
    -rwxrwxr-x  1 www-data www-data 236K Aug 13 10:02 umemcache.so
    drwxrwxr-x  3 www-data www-data 4.0K Aug 13 10:02 unittest2
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 unittest2-0.5.1.dist-info
    drwxrwxr-x  4 www-data www-data 4.0K Aug 13 10:02 venusian
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 venusian-1.0.dist-info
    drwxrwxr-x  3 www-data www-data 4.0K Aug 13 10:02 waitress
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 waitress-0.9.0.dist-info
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 webob
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 WebOb-1.4.1.dist-info
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 webtest
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 WebTest-2.0.23.dist-info
    drwxrwxr-x  5 www-data www-data 4.0K Aug 13 10:00 wheel
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:00 wheel-0.29.0.dist-info
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 wsgiproxy
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 WSGIProxy-0.2.2.dist-info
    drwxrwxr-x  6 www-data www-data 4.0K Aug 13 10:02 zope
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 zope.component-4.2.1.dist-info
    -rw-rw-r--  1 www-data www-data  299 Aug 13 10:02 zope.component-4.2.1-py2.7-nspkg.pth
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 zope.deprecation-4.1.2.dist-info
    -rw-rw-r--  1 www-data www-data  299 Aug 13 10:02 zope.deprecation-4.1.2-py2.7-nspkg.pth
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 zope.event-4.2.0.dist-info
    -rw-rw-r--  1 www-data www-data  299 Aug 13 10:02 zope.event-4.2.0-py2.7-nspkg.pth
    drwxrwxr-x  2 www-data www-data 4.0K Aug 13 10:02 zope.interface-4.2.0.dist-info
    -rw-rw-r--  1 www-data www-data  299 Aug 13 10:02 zope.interface-4.2.0-py2.7-nspkg.pth

**Comment by DigNative** [comment]
I have just figured out that `local/lib/python2.7/site-packages/syncserver.egg-link` contained an absolute path which was not correctly set. This is because I compiled the cloned repository in a different path before moving it to the final location --- the documentation doesn't mention that there are absolute paths set in other files than in the `syncserver.ini` file.

After correcting the path in the file mentioned above, everything seems to work fine right now. I suggest to amend the documentation to contain some words on necessary changes when you move the syncserver directory to another location after compilation.
